File: Configs/.config/quickshell/LooknfeelLua.js
.pragma library

// Renders the managed block the panel writes, then parses the reader records,
// theme variables, and `hyprctl -j --batch getoption` output it reads back.
//
// The rendered string is the same one handed to `hyprctl eval` for the live
// preview, so preview and saved state cannot drift. Keep it that way: a second
// renderer anywhere is a second grammar.

var BEGIN_FENCE = "-- >>> look and feel (generated; edit in the panel) >>>"
var END_FENCE = "-- <<< look and feel <<<"

// ------------------------------------------------------------- rendering

function indent(depth) {
    var pad = ""
    for (var i = 0; i < depth; i++) pad += "    "
    return pad
}

function renderValue(value) {
    if (typeof value === "boolean") return value ? "true" : "false"
    if (typeof value === "number") return String(value)
    return '"' + String(value).replace(/\\/g, "\\\\").replace(/"/g, '\\"') + '"'
}

// Scalars before nested tables, so the generated Lua reads the way a
// hand-written config section does.
function renderTable(node, depth) {
    var scalars = []
    var nested = []
    for (var key in node) {
        if (node[key] !== null && typeof node[key] === "object") nested.push(key)
        else scalars.push(key)
    }
    var lines = []
    var i
    for (i = 0; i < scalars.length; i++)
        lines.push(indent(depth + 1) + scalars[i] + " = " + renderValue(node[scalars[i]]))
    for (i = 0; i < nested.length; i++)
        lines.push(indent(depth + 1) + nested[i] + " = " + renderTable(node[nested[i]], depth + 1))
    return "{\n" + lines.join(",\n") + "\n" + indent(depth) + "}"
}

function buildTree(overrides) {
    var root = {}
    for (var key in overrides) {
        var parts = String(key).split(":")
        var cursor = root
        for (var i = 0; i < parts.length - 1; i++) {
            if (!cursor[parts[i]] || typeof cursor[parts[i]] !== "object")
                cursor[parts[i]] = {}
            cursor = cursor[parts[i]]
        }
        cursor[parts[parts.length - 1]] = overrides[key]
    }
    return root
}

function renderAnimation(animation) {
    var parts = ['leaf = "' + animation.leaf + '"']
    parts.push("enabled = " + (animation.enabled === false ? "false" : "true"))
    if (animation.speed !== undefined && animation.speed !== null && animation.speed !== "")
        parts.push("speed = " + animation.speed)
    if (animation.bezier) parts.push('bezier = "' + animation.bezier + '"')
    if (animation.style) parts.push('style = "' + animation.style + '"')
    return "hl.animation({ " + parts.join(", ") + " })"
}

// Returns "" when there is nothing to write, which is what lets clearing the
// last override delete the file and hand the keys back to the theme layer.
function renderBlock(overrides, animations, variables) {
    var keys = []
    for (var key in (overrides || {})) keys.push(key)
    var list = animations || []
    var variableKeys = []
    for (var variable in (variables || {})) variableKeys.push(variable)
    if (keys.length === 0 && list.length === 0 && variableKeys.length === 0) return ""

    var body = []
    if (variableKeys.length > 0) {
        variableKeys.sort()
        body.push('local vars = require("vars")')
        for (var v = 0; v < variableKeys.length; v++) {
            var name = variableKeys[v]
            // Hypr's vars module stores strings, and the shared shell config
            // parser deliberately accepts only this quoted generated form.
            body.push("vars.set(" + renderValue(name) + ", "
                + renderValue(String(variables[name])) + ")")
        }
    }
    if (keys.length > 0)
        body.push("hl.config(" + renderTable(buildTree(overrides), 0) + ")")
    for (var i = 0; i < list.length; i++)
        body.push(renderAnimation(list[i]))

    return BEGIN_FENCE + "\n" + body.join("\n") + "\n" + END_FENCE + "\n"
}

// --------------------------------------------------------------- parsing

function parseRecords(text) {
    var keys = {}
    var variables = {}
    var animations = []
    var lines = String(text || "").split("\n")
    for (var i = 0; i < lines.length; i++) {
        if (!lines[i]) continue
        var f = lines[i].split("\t")
        if (f[0] === "k") {
            var raw = f[3]
            keys[f[1]] = f[2] === "number" ? Number(raw)
                : f[2] === "boolean" ? raw === "true"
                : raw
        } else if (f[0] === "v") {
            variables[f[1]] = f[3]
        } else if (f[0] === "a") {
            animations.push({
                leaf: f[1],
                enabled: f[2] === "true",
                speed: f[3] === "" || f[3] === undefined ? 0 : Number(f[3]),
                bezier: f[4] || "",
                style: f[5] || ""
            })
        }
    }
    return { keys: keys, variables: variables, animations: animations }
}

// Custom-type values serialize as four edge values ("7 7 7 7"). Nothing in this
// config sets them per-edge, so the first field is the value.
function cssScalar(css) {
    var first = String(css === undefined ? "" : css).trim().split(/\s+/)[0]
    var value = Number(first)
    return isNaN(value) ? 0 : value
}

// `hyprctl -j --batch` emits one chunk per command, blank-line separated, and
// does not abort on a bad key: an inactive layout engine's options come back as
// a bare `no such option` line. Skip anything that is not JSON.
function parseGetoption(text) {
    var out = {}
    var chunks = String(text || "").split(/\n\s*\n/)
    for (var i = 0; i < chunks.length; i++) {
        var chunk = chunks[i].trim()
        if (!chunk || chunk.charAt(0) !== "{") continue
        var data
        try {
            data = JSON.parse(chunk)
        } catch (error) {
            continue
        }
        if (!data || !data.option) continue

        var entry = null
        if ("int" in data) entry = { value: data.int, type: "int" }
        else if ("float" in data) entry = { value: data.float, type: "float" }
        else if ("bool" in data) entry = { value: data.bool, type: "bool" }
        else if ("str" in data) entry = { value: data.str, type: "str" }
        else if ("css" in data) entry = { value: cssScalar(data.css), type: "css" }
        if (!entry) continue

        entry.set = data.set === true
        out[data.option] = entry
    }
    return out
}

// The theme pack's own values, straight from themes/theme.lua. Unlike a
// getoption read these are not clouded by the panel's override block, so a row
// dialled back to the theme's value can be recognised as no longer overridden.
function parseThemeConfig(text) {
    var out = {}
    var rx = /runtime\.config\(\s*"([^"]+)"\s*,\s*(.+?)\s*\)\s*$/
    var lines = String(text).split("\n")
    for (var i = 0; i < lines.length; i++) {
        var match = rx.exec(lines[i].trim())
        if (!match) continue
        var value = match[2]
        if (value.charAt(0) === '"') value = value.slice(1, -1)
        else if (value === "true" || value === "false") value = value === "true"
        else if (!isNaN(Number(value))) value = Number(value)
        out[match[1].split(".").join(":")] = value
    }
    return out
}

// Theme variables need their own namespace: CURSOR_SIZE is not a Hyprland
// option and must never be mistaken for one by baseline/reset handling.
function parseThemeVariables(text) {
    var out = {}
    var rx = /vars\.set\(\s*"([^"]+)"\s*,\s*"([^"]*)"\s*\)\s*$/
    var lines = String(text).split("\n")
    for (var i = 0; i < lines.length; i++) {
        var match = rx.exec(lines[i].trim())
        if (match) out[match[1]] = match[2]
    }
    return out
}

File: Configs/.local/lib/hypr/window/lib/looknfeel_lua.py
"""Renders the managed block the TUI writes, then parses the reader records,
theme variables, and `hyprctl -j --batch getoption` output it reads back.

The rendered string is the same one handed to `hyprctl eval` for the live
preview, so preview and saved state cannot drift. Keep it that way: a second
renderer anywhere is a second grammar.
"""

import json
import re

BEGIN_FENCE = "-- >>> look and feel (generated; edit in the panel) >>>"
END_FENCE = "-- <<< look and feel <<<"


def _indent(depth):
    return "    " * depth


def _number(value):
    """Lua accepts both, but an integral float must render as `7` and not `7.0`
    so the block is byte-identical to what the previous renderer emitted and to
    what the fixtures expect."""
    if isinstance(value, float) and value.is_integer():
        return str(int(value))
    return str(value)


def render_value(value):
    if isinstance(value, bool):
        return "true" if value else "false"
    if isinstance(value, (int, float)):
        return _number(value)
    return '"' + str(value).replace("\\", "\\\\").replace('"', '\\"') + '"'


def render_table(node, depth):
    """Scalars before nested tables, so the generated Lua reads the way a
    hand-written config section does."""
    scalars = [k for k, v in node.items() if not isinstance(v, dict)]
    nested = [k for k, v in node.items() if isinstance(v, dict)]

    lines = [_indent(depth + 1) + k + " = " + render_value(node[k]) for k in scalars]
    lines += [_indent(depth + 1) + k + " = " + render_table(node[k], depth + 1)
              for k in nested]
    return "{\n" + ",\n".join(lines) + "\n" + _indent(depth) + "}"


def build_tree(overrides):
    root = {}
    for key, value in overrides.items():
        parts = str(key).split(":")
        cursor = root
        for part in parts[:-1]:
            if not isinstance(cursor.get(part), dict):
                cursor[part] = {}
            cursor = cursor[part]
        cursor[parts[-1]] = value
    return root


def render_animation(animation):
    parts = ['leaf = "%s"' % animation.get("leaf", "")]
    parts.append("enabled = " + ("false" if animation.get("enabled") is False else "true"))
    speed = animation.get("speed")
    if speed is not None and speed != "":
        parts.append("speed = " + _number(speed))
    if animation.get("bezier"):
        parts.append('bezier = "%s"' % animation["bezier"])
    if animation.get("style"):
        parts.append('style = "%s"' % animation["style"])
    return "hl.animation({ " + ", ".join(parts) + " })"


def render_block(overrides, animations=None, variables=None):
    """Returns "" when there is nothing to write, which is what lets clearing
    the last override delete the file and hand the keys back to the theme
    layer."""
    overrides = overrides or {}
    animations = animations or []
    variables = variables or {}
    if not overrides and not animations and not variables:
        return ""

    body = []
    if variables:
        body.append('local vars = require("vars")')
        for name in sorted(variables):
            # Hypr's vars module stores strings, and the shared shell config
            # parser deliberately accepts only this quoted generated form.
            body.append("vars.set(%s, %s)"
                        % (render_value(name), render_value(str(variables[name]))))
    if overrides:
        body.append("hl.config(" + render_table(build_tree(overrides), 0) + ")")
    body += [render_animation(a) for a in animations]

    return BEGIN_FENCE + "\n" + "\n".join(body) + "\n" + END_FENCE + "\n"


def _field(fields, index):
    return fields[index] if index < len(fields) else ""


def parse_records(text):
    keys = {}
    variables = {}
    animations = []
    for line in str(text or "").split("\n"):
        if not line:
            continue
        f = line.split("\t")
        kind = f[0]
        if kind == "k":
            raw = _field(f, 3)
            kind_name = _field(f, 2)
            if kind_name == "number":
                keys[f[1]] = _as_number(raw)
            elif kind_name == "boolean":
                keys[f[1]] = raw == "true"
            else:
                keys[f[1]] = raw
        elif kind == "v":
            variables[f[1]] = _field(f, 3)
        elif kind == "a":
            speed = _field(f, 3)
            animations.append({
                "leaf": _field(f, 1),
                "enabled": _field(f, 2) == "true",
                "speed": 0 if speed == "" else _as_number(speed),
                "bezier": _field(f, 4),
                "style": _field(f, 5),
            })
    return {"keys": keys, "variables": variables, "animations": animations}


def _as_number(raw):
    try:
        value = float(raw)
    except (TypeError, ValueError):
        return 0
    return int(value) if value.is_integer() else value


def css_scalar(css):
    """Custom-type values serialize as four edge values ("7 7 7 7"). Nothing in
    this config sets them per-edge, so the first field is the value."""
    parts = str("" if css is None else css).strip().split()
    if not parts:
        return 0
    try:
        value = float(parts[0])
    except ValueError:
        return 0
    return int(value) if value.is_integer() else value


def parse_getoption(text):
    """`hyprctl -j --batch` emits one chunk per command, blank-line separated,
    and does not abort on a bad key: an inactive layout engine's options come
    back as a bare `no such option` line. Skip anything that is not JSON."""
    out = {}
    for chunk in re.split(r"\n\s*\n", str(text or "")):
        chunk = chunk.strip()
        if not chunk or chunk[0] != "{":
            continue
        try:
            data = json.loads(chunk)
        except ValueError:
            continue
        if not isinstance(data, dict) or "option" not in data:
            continue

        entry = None
        for field, name in (("int", "int"), ("float", "float"),
                            ("bool", "bool"), ("str", "str")):
            if field in data:
                entry = {"value": data[field], "type": name}
                break
        if entry is None and "css" in data:
            entry = {"value": css_scalar(data["css"]), "type": "css"}
        if entry is None:
            continue

        entry["set"] = data.get("set") is True
        out[data["option"]] = entry
    return out


_CONFIG_RX = re.compile(r'runtime\.config\(\s*"([^"]+)"\s*,\s*(.+?)\s*\)$')
_VARIABLE_RX = re.compile(r'vars\.set\(\s*"([^"]+)"\s*,\s*"([^"]*)"\s*\)$')


def parse_theme_config(text):
    """The theme pack's own values, straight from themes/theme.lua. Unlike a
    getoption read these are not clouded by the override block, so a row dialled
    back to the theme's value can be recognised as no longer overridden."""
    out = {}
    for line in str(text).split("\n"):
        match = _CONFIG_RX.search(line.strip())
        if not match:
            continue
        raw = match.group(2)
        if raw.startswith('"'):
            value = raw[1:-1]
        elif raw in ("true", "false"):
            value = raw == "true"
        else:
            try:
                number = float(raw)
            except ValueError:
                value = raw
            else:
                value = int(number) if number.is_integer() else number
        out[match.group(1).replace(".", ":")] = value
    return out


def parse_theme_variables(text):
    """Theme variables need their own namespace: CURSOR_SIZE is not a Hyprland
    option and must never be mistaken for one by baseline/reset handling."""
    out = {}
    for line in str(text).split("\n"):
        match = _VARIABLE_RX.search(line.strip())
        if match:
            out[match.group(1)] = match.group(2)
    return out
